### ISSUE TITLE
Gradient 0 with jax.scipy.ndimage.map_coordinates

### ISSUE THREAD

**Issue by anorak94** [google/jax#8684]
I am trying to do a simple rigid registration using jax. My code for doing this is - 

```
def rigid_imx_loss( tx):
    w, h = p1.shape
    origin = (w//2, h//2)
    x0 = jnp.arange(0, w, 1)
    x1 = jnp.arange(0, h, 1)
    X0, X1 = jnp.meshgrid(x0, x1)
    pts = jnp.asarray([X1.ravel(), X0.ravel()]).T
    translate = jnp.asarray([0, tx])
    pts_tx = pts + translate.T
    out = map_coordinates(p1, pts_tx.T, order = 1).reshape(w, h)
    return jnp.mean((out-p2)**2)
```

Getting the grad by doing 
`tx_grad = grad(rigid_imx_loss)(tx)` always returns 0

When I plot the two images the translation part of the code seems to be working. i.e. the image does appear shifted compared to the original image. However my gradient wrt to tx is always 0. I tested it over a range of tx values and while the loss seems to be different the gradient remains 0 regardless.  

<img width="434" alt="Screenshot 2021-11-24 at 19 06 33" src="https://user-images.githubusercontent.com/45297533/143292000-74ff1196-388d-46dc-b654-444b85cfbe77.png">

**Comment by clemisch** [comment]
Not related to the gradient problem, but fyi: Optimizing the shift wrt. the L2 norm is biased towards half-pixel shifts ([source](https://arxiv.org/pdf/1809.05583.pdf), not mine ^^), because the interpolation reduces noise. I think you see this in you cost function plateauing at half-pixel shifts.

**Comment by shoyer** [comment]
Thanks for the report! Could you kindly share the source code for a complete example that I could run to reproduce the issue?

**Comment by anorak94** [comment]
Hey im sorry i think i fixed the issue the images i was passing into the grad function were `uint32` type when i change the type to `float32` it seems to work i.e. I don't get zero-gradients. Feel free to close if this solution seems correct.

**Comment by shoyer** [comment]
Well, linear interpolation on integer valued arrays should still produce continuous valued outputs, so probably there should be non-zero gradients with respect to the coordinares. An example would still be helpful!

**Comment by anorak94** [comment]
`
import jax
import jax.numpy as jnp
from jax import grad
import numpy as np
from jax.scipy.ndimage import map_coordinates

p1 = np.random.randint(low = 0, high = 255, size =(256, 256))
p2 = np.random.randint(low = 0, high = 255, size =(256, 256))

def rigid_imx_loss( tx):
    w, h = p1.shape
    origin = (w//2, h//2)
    x0 = jnp.arange(0, w, 1)
    x1 = jnp.arange(0, h, 1)
    X0, X1 = jnp.meshgrid(x0, x1)
    pts = jnp.asarray([X1.ravel(), X0.ravel()]).T
    translate = jnp.asarray([0, tx])
    pts_tx = pts + translate.T
    out = map_coordinates(p1, pts_tx.T, order = 1).reshape(w, h)
    return jnp.mean((out-p2)**2)

tx = 10.

grad(rigid_imx_loss)(tx)

p1 = p1.astype("float32")
p2 = p2.astype("float32")

grad(rigid_imx_loss)(tx)0
`

This reproduces the error

**Comment by tetterl** [comment]
> Well, linear interpolation on integer valued arrays should still produce continuous valued outputs, so probably there should be non-zero gradients with respect to the coordinares. An example would still be helpful!

`jax.scipy.ndimage.map_coordinates` outputs integer values for integer inputs which explains the zero gradient. Unfortunately there doesn't seem to be a way to omit the rounding to integers.
